src/greatwall/resources/greatwall.py

```diff
@@ -234,3 +268,99 @@ def return_level(self):
             self.is_finished = False
         self.current_level -= 1
         self.state = self.protocol_states[self.current_level]
+
+        self.history_step_back()
+        self.history_state_load()
+
+    def history_reset(self):
+
+        self.saved_fractals = {}
+        self.saved_path = []
+        self.saved_states = {}
+
+    # maps a tree path to an index key for the "self.saved_states" dictionary
+    #
+    # EXAMPLES:
+    # the path [] maps to the index 0, for any tree
+    # the path [1, 2] maps to the index 23, for a tree of arity 3 and depth 4
+    # the path [0, 1, 1] maps to the index 13, for a tree of arity 2 and depth 5
+    def history_path_to_index(self, path):
+
+        def calculate_tree_size(depth):
+            size = 0
+            span = 1
+            while depth > 0:
+                depth = depth - 1
+                size = size + span
+                span = span * self.tree_arity
+            return size
+
+        index = 0
+        depth = self.tree_depth
+        for position in range(0, len(path)):
+            index = index + 1
+            depth = depth - 1
+
+            choice = path[position]
+
+            tree_size = calculate_tree_size(depth)
+            index = index + (tree_size * choice)
+
+        return index
+
+    def history_step_back(self):
+
+        if len(self.saved_path) == 0:
+            return
+
+        self.saved_path.pop()
+
+    def history_step_next(self, index):
```

Comment:
Because these methods, `history_step_next` and `history_step_back`, have the similar names, the end user will expecting the same returned values from them. I propose using boolean values as returned values from them instead of `None` in `history_step_back`.

Reply:
This is a discussion purely about style. The code is readable and works as it is, there is no need for change here.